### ISSUE TITLE
Memory leak

### ISSUE THREAD

**Issue by fnadeau** [cgreen-devs/cgreen#260]
From master (9bcbaae7), running ```make valgrind``` return this:
```
grep --with-filename --line-number " lost:" valgrind.log | grep -v " 0 bytes"
valgrind.log:1374:==19258==    definitely lost: 8 bytes in 1 blocks
valgrind.log:1414:==19259==    definitely lost: 16 bytes in 2 blocks
valgrind.log:4930:==19521==    definitely lost: 8 bytes in 1 blocks
valgrind.log:4970:==19522==    definitely lost: 16 bytes in 2 blocks
valgrind.log:10151:==19870==    definitely lost: 8 bytes in 1 blocks
valgrind.log:10191:==19871==    definitely lost: 16 bytes in 2 blocks
```
The following seems to fix
```
diff --git a/src/constraint.c b/src/constraint.c
index 723d8de..d464c27 100644
--- a/src/constraint.c
+++ b/src/constraint.c
@@ -464,6 +464,7 @@ Constraint *create_return_by_value_constraint(intptr_t value_to_return, size_t s
     constraint->type = CGREEN_RETURN_VALUE_CONSTRAINT;
 
     constraint->compare = &compare_true;
+    constraint->destroy = &destroy_string_constraint;
     constraint->execute = &test_true;
     constraint->name = "return value";
     constraint->expected_value = make_cgreen_by_value((void*)actual_return, size);
```

I'm not sure if it's a proper fix, it seems like a hack. Also, makefile gives an error:
```
[100%] Built target cgreen_runner_tests
Running all tests under Valgrind ..........................
grep --with-filename --line-number " lost:" valgrind.log | grep -v " 0 bytes"
make: *** [Makefile:135: valgrind] Error 1
```
So it ain't a complete fix.

edit: fix return value of ```make valgind```, old behavior was to return 6, (wc -l)

**Comment by thoni56** [comment]
Yes, I discovered this too, while commenting on your PR. So I changed the `valgrind` target to better show the loss.

Your fix does not work which `make unit` will show.

The error you see when `make valgrind` after your attempt to fix is, I think, because there are memory violations while running under valgrind. (I need to verify this...)

The problem is probably that the area that is lost is allocated in `create_by_value_constraint()` to hold the "by value" value. If there is an error, then that area might be referenced by the text reporter. And if it's returned already when tallying the mocks then that areas is no longer accessible.

I have just discovered this, so I'm not 100% sure. E.g. wouldn't the string constraints suffer from the same problem...

**Comment by thoni56** [comment]
Your suggested fix might feel like a hack because it uses the existing `destroy_string_constraint()` function for a "by value" constraint. But it is not really wrong. I did the same thing but implemented a new `destroy_by_value_constraint()` that does the exact same thing.

But then ` make unit` gives these errors:

```
Running "libmocks_tests" (40 tests)...
cgreen/tests/mocks_tests.c:459: Failure: Mocks -> can_return_by_value
        Expected [actualBox.weight] to [equal] [20]
                actual value:                   [0]
                expected value:                 [20]

cgreen/tests/mocks_tests.c:460: Failure: Mocks -> can_return_by_value
        Expected [actualBox.height] to [equal] [10]
                actual value:                   [0]
                expected value:                 [10]

cgreen/tests/mocks_tests.c:475: Failure: Mocks -> can_return_by_value_depending_on_input_parameter
        Expected [retrievedBox1.weight] to [equal] [20]
                actual value:                   [21935]
                expected value:                 [20]

cgreen/tests/mocks_tests.c:476: Failure: Mocks -> can_return_by_value_depending_on_input_parameter
        Expected [retrievedBox1.height] to [equal] [10]
                actual value:                   [-2045952624]
                expected value:                 [10]

cgreen/tests/mocks_tests.c:478: Failure: Mocks -> can_return_by_value_depending_on_input_parameter
        Expected [retrievedBox2.weight] to [equal] [33]
                actual value:                   [21935]
                expected value:                 [33]

cgreen/tests/mocks_tests.c:479: Failure: Mocks -> can_return_by_value_depending_on_input_parameter
        Expected [retrievedBox2.height] to [equal] [5]
                actual value:                   [-2045955248]
                expected value:                 [5]

  "Mocks": 81 passes, 6 failures in 7ms.
Completed "libmocks_tests": 81 passes, 6 failures in 7ms.
```

So I'll have to investigate further.

**Comment by thoni56** [comment]
Yes! This was what I was looking for:

```
==10846== Invalid read of size 8
==10846==    at 0x48685A2: retrieveBox (mocks_tests.c:450)
==10846==    by 0x486861D: Mocks__can_return_by_value (mocks_tests.c:458)
==10846==    by 0x4856055: run (runner.c:272)
==10846==    by 0x48561E4: run_the_test_code (runner.c:325)
==10846==    by 0x48586F6: run_test_in_its_own_process (posix_runner_platform.c:41)
==10846==    by 0x4856372: run_every_test (runner.c:85)
==10846==    by 0x485632A: run_every_test (runner.c:68)
==10846==    by 0x4856574: run_test_suite (runner.c:39)
==10846==    by 0x10B8BE: run_tests (runner.c:225)
==10846==    by 0x10BC5F: runner (runner.c:306)
==10846==    by 0x10A811: run_tests_in_library (cgreen-runner.c:168)
==10846==    by 0x10AE0F: main (cgreen-runner.c:312)
==10846==  Address 0x4bf3940 is 0 bytes inside a block of size 8 free'd
==10846==    at 0x483CA3F: free (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==10846==    by 0x4850EAB: destroy_cgreen_value (cgreen_value.c:54)
==10846==    by 0x485146B: destroy_by_value_constraint (constraint.c:731)
==10846==    by 0x48515A6: destroy_constraint (constraint.c:132)
==10846==    by 0x48575CD: destroy_cgreen_vector (vector.c:34)
==10846==    by 0x485438F: destroy_expectation (mocks.c:710)
==10846==    by 0x485495F: destroy_expectation_if_time_to_die (mocks.c:420)
==10846==    by 0x4854EEC: mock_ (mocks.c:279)
==10846==    by 0x48685A1: retrieveBox (mocks_tests.c:450)
==10846==    by 0x486861D: Mocks__can_return_by_value (mocks_tests.c:458)
==10846==    by 0x4856055: run (runner.c:272)
==10846==    by 0x48561E4: run_the_test_code (runner.c:325)
==10846==  Block was alloc'd at
==10846==    at 0x483B7F3: malloc (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==10846==    by 0x485242C: create_return_by_value_constraint (constraint.c:461)
==10846==    by 0x48685E0: Mocks__can_return_by_value (mocks_tests.c:455)
==10846==    by 0x4856055: run (runner.c:272)
==10846==    by 0x48561E4: run_the_test_code (runner.c:325)
==10846==    by 0x48586F6: run_test_in_its_own_process (posix_runner_platform.c:41)
==10846==    by 0x4856372: run_every_test (runner.c:85)
==10846==    by 0x485632A: run_every_test (runner.c:68)
==10846==    by 0x4856574: run_test_suite (runner.c:39)
==10846==    by 0x10B8BE: run_tests (runner.c:225)
==10846==    by 0x10BC5F: runner (runner.c:306)
==10846==    by 0x10A811: run_tests_in_library (cgreen-runner.c:168)
```

So I'm on the right track.

**Comment by thoni56** [comment]
Fixed by recent commits ending with fcaf37aba8d92b1e6c7634fdba61769b4152f722.
